File: php/array_merge.php
<?php

$a0 = array();
$a1 = array("red", "green");
$a2 = array("red", "blue"); // repeated element "red"

$m0 = array();
$m1 = array(
		1000 => array(
				"a11" => 111,
				"a12" => 112
		),
		1001 => array(
				"b11" => 121,
				"b12" => 122
		),
		"asdf" => array(
				"asdf11" => 131,
				"asdf12" => 132
		)
);
$m2 = array(
		1001 => array( // repeated key
				"b11" => 121, // repeated key, repeated value
				"b12" => 222 // repeated key, different value
		),
		"asdf" => array( // repeated key
				"asdf11" => 131, // repeated key, repeated value
				"asdf12" => 232 // repeated key, different value
		),
		1002 => array(
				"c11" => 241,
				"c12" => 242
		),
);



echo "<pre>";

echo "\n\n\narray_merge(a1, a2)\n\n";
$merge1 = array_merge($a1, $a2);
var_dump($merge1);

echo "\n\n\narray_merge(a0, a1)\n\n";
$merge2 = array_merge($a0, $a1);
var_dump($merge2);

echo "\n\n\narray_merge(m1, m2)\n\n";
$merge3 = array_merge($m1, $m2);
var_dump($merge3);
echo "\n";
echo "numeric keys are NOT preserved.\n";
echo "numeric keys are NOT merged.\n";
echo "string keys of the second dimension are overwritten.\n";

echo "\n\n\narray_merge_recursive(m1, m2)\n\n";
$merge4 = array_merge_recursive($m1, $m2);
var_dump($merge4);
echo "\n";
echo "numeric keys are NOT preserved.\n";
echo "numeric keys are NOT merged.\n";
echo "string keys of the second dimension are merged.\n";

echo "</pre>";

if (!class_exists('PHPUnit_Framework_TestCase')) exit;

class Test extends PHPUnit_Framework_TestCase
{
  public function test()
  {
    global $merge1;
    $this->assertEquals(["red", "green", "red", "blue"], $merge1);
  }

  public function test2()
  {
    global $merge2;
    $this->assertEquals($merge2, ["red", "green"]);
  }

  public function test3()
  {
    global $merge3;
    $this->assertEquals($merge3, [
        0 => [
          "a11" => 111, 
          "a12" => 112], 
        1 => [
          "b11" => 121, 
          "b12" => 122], 
        "asdf" => [
          "asdf11" => 131, 
          "asdf12" => 232],
        2 => [
          "b11" => 121, 
          "b12" => 222],
        3 => [
          "c11" => 241, 
          "c12" => 242]
    ]);
  }

  public function test4()
  {
    global $merge4;
    $this->assertEquals($merge4, [
        0 => [
          "a11" => 111, 
          "a12" => 112], 
        1 => [
          "b11" => 121, 
          "b12" => 122], 
        "asdf" => [
          "asdf11" => [
            131, 131], 
          "asdf12" => [
            132, 232]],
        2 => [
          "b11" => 121, 
          "b12" => 222],
        3 => [
          "c11" => 241, 
          "c12" => 242]
    ]);
  }

}

?>

File: python/array_merge.py
import pprint
import unittest

a0 = []
a1 = ["red", "green"]
a2 = ["red", "blue"]

m0 = []
m1 = {
    1000: {
      'a11': 111,
      'a12': 112
    },
    1001: {
      'b11': 121,
      'b12': 122
    },
    'asdf': {
      'asdf11': 131,
      'asdf12': 132
    }
}
m2 = {
    1001: { # repeated key
      'b11': 121, # repeated key, repeated value
      'b12': 222 # repeated key, different value
    },
    'asdf': {
      'asdf11': 131, # repeated key, repeated value
      'asdf12': 232 # repeated key, different value
    },
    1002: {
      'c11': 241,
      'c12': 242
    }
}



def mergeArrays():
    return a1 + a2

def mergeArrayNull():
    return a0 + a1

def mergeDicts():
    merge = m1.copy()
    merge.update(m2)
    return merge

def recMergeDicts():
    merge = m1.copy()
    merge = dict_merge_recursive(merge, m2)
    return merge

# wsgi header
def application(environ, start_response):
    status = '200 OK'
    o = '' # output



    o += "<pre>"

    o += "\n\n\n\na1 + a2\n\n"
    merge = mergeArrays()
    o += pprint.pformat(merge)

    o += "\n\n\n\na0 + a1\n\n"
    merge = mergeArrayNull()
    o += pprint.pformat(merge)

    o += "\n\n\n\nm1.copy(), .update(m2)\n\n"
    merge = mergeDicts()
    o += pprint.pformat(merge)
    o += "\nnumeric keys are preserved"
    o += "\nexisting keys are overwritten"

    o += "\n\n\n\nrecursive merge\n\n"
    merge = recMergeDicts()
    o += pprint.pformat(merge)
    o += "\nnumeric keys are preserved"
    o += "\nexisting keys are merged"

    o += "</pre>"



# wsgi footer
    response_headers = [('Content-type', 'text/html')]
    start_response(status, response_headers)
    return [o]

def dict_merge_recursive(d1, d2):
    res = {}

    if (type(d1) != dict or
            type(d2) != dict):
        return [d1, d2]

    for p in d1:
        res[p] = d1[p]

    for p in d2:
        if p in res:
            res[p] = dict_merge_recursive(res[p], d2[p])
        else:
            res[p] = d2[p]

    return res



class UnitTests(unittest.TestCase):
  def testA(self):
    self.assertEqual(mergeArrays(), ['red', 'green', 'red', 'blue'])

  def testB(self):      
    self.assertEqual(mergeArrayNull(), ['red', 'green'])

  def testC(self):
    self.assertEqual(mergeDicts(), {
        1000: {'a11': 111, 'a12': 112},
        1001: {'b11': 121, 'b12': 222},
        1002: {'c11': 241, 'c12': 242},
        'asdf': {'asdf11': 131, 'asdf12': 232}
    })

  def testD(self):
    self.assertEqual(recMergeDicts(), {
        1000: {'a11': 111, 'a12': 112},
        1001: {'b11': [121, 121], 'b12': [122, 222]},
        1002: {'c11': 241, 'c12': 242},
        'asdf': {'asdf11': [131, 131], 'asdf12': [132, 232]}
    })

if __name__  == '__main__':
  unittest.main()

